SSAO delayed resize fix

GSterbrant opened the pull request:
Fixes #4239 

Removed the resize timeouts. Added a general resize callback to the post effects such that they can handle resize. 

Bloom needs such a callback to adapt it's internal targets (which should save performance when viewport is small).

mvaligursky commented on src/framework/components/camera/post-effect-queue.js:
```diff
@@ -373,26 +365,10 @@ class PostEffectQueue {
-            this.resizeTimeout = null;
-        }
 
         this.resizeLast = now();
```
should this line be removed?

GSterbrant replied:
Ah, thanks!

slimbuck commented on scripts/posteffects/posteffect-ssao.js:
```diff
@@ -17,23 +17,25 @@ function SSAOEffect(graphicsDevice, ssaoScript) {
     this.ssaoScript = ssaoScript;
     this.needsDepthBuffer = true;
 
+    var vshader = [
+        (graphicsDevice.webgl2) ? ("#version 300 es\n\n" + pc.shaderChunks.gles3VS) : "",
+        graphicsDevice.webgl2 ? "" : "#extension GL_OES_standard_derivatives: enable\n",
+        graphicsDevice.webgl2 ? "" : "attribute vec2 aPosition;",
+        "",
+        "varying vec2 vUv0;",
+        "",
+        "void main(void)",
+        "{",
+        graphicsDevice.webgl2 ? "    vUv0 = vec2((gl_VertexID << 1) & 2, gl_VertexID & 2);" : "   vUv0 = (aPosition.xy + 1.0) * 0.5;",
+        graphicsDevice.webgl2 ? "    gl_Position = vec4(vUv0 * 2.0f - 1.0f, 0.0f, 1.0f);" : "    gl_Position = vec4(aPosition, 0.0, 1.0);",
+        "}"
+    ].join("\n");
+
```
I feel like this would be simpler as two different shaders?

GSterbrant replied:
Ah, oh my. Thanks for catching this, I think this shader won't work.

mvaligursky replied:
Or just use webgl1 version for both platforms? Not sure we win much with custom webgl2 shader

slimbuck replied:
Yeah I have no idea why the fancy bit ops on webgl2. The headers are still different though between 1 and 2.

GSterbrant replied:
This was a part of another piece of work I did and accidentally carried over in this PR.

mvaligursky commented on src/framework/components/camera/post-effect-queue.js:
```diff
@@ -404,6 +380,8 @@ class PostEffectQueue {
         for (let i = 0, len = effects.length; i < len; i++) {
             const fx = effects[i];
+            if (fx.effect.resize !== undefined)
+                fx.effect.resize();
```
should this be called after _resizeOffscreenTarget so that the effect can use the siez of those targets for resizing?

GSterbrant replied:
Is it better that they depend on the input target or the graphics device size? The input target could be quarter res but the depth target full res, for example.

mvaligursky replied:
what if we render with camera to a different size or even different aspect than device size? Device size should not be used at all.

GSterbrant replied:
True. Okay I'll change this to be dependent on the input target.

GSterbrant replied:
Okay, so looking back at this, every post effect is created without the input target and rely on the device size on creation. I think considering these parameters, the base size of the effect shouldn't be handled by the effect at all, but rather by the post effects queue.

So on init, we should also pass a context containing the size for the effect. I guess we can add that to the list of things we want to do when refactoring the post effects stuff 🙂 .

mvaligursky commented on src/framework/components/camera/post-effect-queue.js:
```diff
@@ -404,6 +368,8 @@ class PostEffectQueue {
 
         for (let i = 0, len = effects.length; i < len; i++) {
             const fx = effects[i];
+            if (fx.effect.resize !== undefined)
```
```suggestion
            if (fx.effect.resize)
```

mvaligursky commented on scripts/posteffects/posteffect-bloom.js:
```diff
@@ -171,12 +171,47 @@ function BloomEffect(graphicsDevice) {
+BloomEffect.prototype.resize = function (target) {
+
+    var width, height;
+    if (target == null) {
```
```suggestion
    if (target === null) {
```

mvaligursky commented on scripts/posteffects/posteffect-bloom.js:
```diff
@@ -171,12 +171,47 @@ function BloomEffect(graphicsDevice) {
+        height = target.colorBuffer.height;
+    }
+
+    if (width == this.width && height == this.height)
```
```suggestion
    if (width === this.width && height === this.height)
```

mvaligursky commented on scripts/posteffects/posteffect-bloom.js:
```diff
@@ -171,12 +171,47 @@ function BloomEffect(graphicsDevice) {
+        return;
+
+    if (this.targets) {
+        for (let i = 0; i < this.targets.length; i++) {
```
cannot use `let` in ES5, only `var`

mvaligursky commented on scripts/posteffects/posteffect-ssao.js:
```diff
@@ -400,16 +378,22 @@ function SSAOEffect(graphicsDevice, ssaoScript) {
+    }
 
     // If no change, skip resize
     if (width == this.width && height == this.height)
```
could you please use `===` here as well

mvaligursky reviewed (approved):
approving with some minor comments (please fix before the merge)

lucaheft commented:
Also the new resize function should be added to the pc.PostEffect docs.
(https://developer.playcanvas.com/en/api/pc.PostEffect.html)

mvaligursky commented:
Perhaps the resize function should be called internally within the effect, and not from the engine (post-effect-queue). Inside the render call, the effect should check the sizes of its internal buffers vs the source / target buffer it has, and rescale those internal buffers as needed.

GSterbrant commented:
The effect itself doesn't know about the input target and has to defer to the graphics device for dimensions. The post effects queue knows about the input targets, and can therefore forward it to the effect when it's enabled/disabled. 

Wouldn't it be better if we can avoid dealing with resize related code if we're not resizing?

mvaligursky commented:
the render function of the effect knows input and output
`    render: function (inputTarget, outputTarget, rect)`

handling size compare is not very expensive, we handle it in few places like this. It's simpler for the user / creator of the effect.

https://github.com/playcanvas/engine/blob/24f325f078c2753ae15948450d3da4b0918932a6/src/framework/scene-grab.js#L156-L158

mvaligursky commented:
This way you also avoid to maintain / handle resizing while the effect / camera is disabled and similar. Each time you render, you make sure targets are the right size.

lucaheft commented:
I like that approach. This way the resizing can be handled on a per post-effect instance.
The post-effect-queue shouldnt handle it anyway, it's just a simple queue isn't it?

lucaheft commented:
Also what i noticed, in the post-effect-queue the width and height is calculated like this:
https://github.com/playcanvas/engine/blob/24f325f078c2753ae15948450d3da4b0918932a6/src/framework/components/camera/post-effect-queue.js#L364-L365
In the bloom post effect like this:
https://github.com/playcanvas/engine/blob/24f325f078c2753ae15948450d3da4b0918932a6/scripts/posteffects/posteffect-bloom.js#L193-L194
And in ssao like this:
https://github.com/playcanvas/engine/blob/24f325f078c2753ae15948450d3da4b0918932a6/scripts/posteffects/posteffect-ssao.js#L391-L392
Shouldnt those calculations all be the same? Except for the downscale in the ssao effect.

GSterbrant commented:
The issue of having to make this calculation at all is due to the fact that post effects are setup prior to knowing the input target. This will be fixed in the PR I am working on addressing these issues :).

The post effect queue calculates the size of the internal targets, and really the effects should just read those values. The device dimensions shouldn't be used at all.

GSterbrant commented:
https://github.com/playcanvas/engine/pull/4397
